File: backend/app/services/agents/tool_events.py
from __future__ import annotations

import json
from typing import Any

from sqlalchemy.orm import Session

from app.models import ToolInvocation

FAILED_STATUSES = {"failed", "error", "cancelled", "timeout"}


def tool_event_from_record(db: Session, record: dict[str, Any]) -> dict[str, Any]:
    """把一次 Function Call 执行记录压缩成可展示、可持久化的工具事件。"""
    result = record.get("result") if isinstance(record.get("result"), dict) else {}
    output = _output_value(result)
    invocation = _tool_invocation(db, result.get("invocation_id"))
    status = _status(record, result, invocation)
    event: dict[str, Any] = {
        "tool_name": str(record.get("tool_name") or result.get("tool_name") or ""),
        "tool_call_id": str(record.get("tool_call_id") or ""),
        "status": status,
    }
    if invocation:
        event["invocation_id"] = invocation.id
        event["duration_ms"] = invocation.duration_ms
        event["status"] = invocation.status or status
        if invocation.error_message:
            event["error"] = _short(invocation.error_message)
    _merge_output_fields(event, output)
    event["summary"] = _summary(event, output)
    return {key: value for key, value in event.items() if value not in (None, "")}


def _tool_invocation(db: Session, invocation_id: Any) -> ToolInvocation | None:
    if not invocation_id:
        return None
    try:
        return db.get(ToolInvocation, str(invocation_id))
    except Exception:
        return None


def _status(
    record: dict[str, Any],
    result: dict[str, Any],
    invocation: ToolInvocation | None,
) -> str:
    if invocation and invocation.status:
        return str(invocation.status)
    value = record.get("status") or result.get("status") or "unknown"
    return str(value)


def _output_value(result: dict[str, Any]) -> Any:
    if "output" in result:
        return result.get("output")
    if "result" in result:
        return result.get("result")
    return result


def _merge_output_fields(event: dict[str, Any], output: Any) -> None:
    if not isinstance(output, dict):
        if output:
            event["stdout"] = _short(output)
        return
    nested = output.get("result") if isinstance(output.get("result"), dict) else {}
    event["exit_code"] = _first(output, "exit_code", "return_code", "code") or _first(
        nested,
        "exit_code",
        "return_code",
        "code",
    )
    event["duration_ms"] = event.get("duration_ms") or _first(
        output,
        "duration_ms",
        "elapsed_ms",
    ) or _first(nested, "duration_ms", "elapsed_ms")
    event["stdout"] = _short(
        _first(output, "stdout", "stdout_tail", "output", "text")
        or _first(nested, "stdout", "stdout_tail", "output", "text")
    )
    event["stderr"] = _short(
        _first(output, "stderr", "stderr_tail") or _first(nested, "stderr", "stderr_tail")
    )
    event["error"] = _short(
        _first(output, "error", "error_message") or _first(nested, "error", "error_message")
    )


def _first(data: dict[str, Any], *keys: str) -> Any:
    for key in keys:
        value = data.get(key)
        if value not in (None, ""):
            return value
    return None


def _summary(event: dict[str, Any], output: Any) -> str:
    if str(event.get("status", "")).lower() in FAILED_STATUSES:
        return _short(event.get("error") or event.get("stderr") or output) or "工具调用失败"
    return _short(event.get("stdout") or output) or "工具调用完成"


def _short(value: Any, limit: int = 500) -> str:
    if value is None:
        return ""
    if isinstance(value, str):
        text = value
    else:
        text = json.dumps(value, ensure_ascii=False, default=str)
    text = " ".join(text.split())
    return text if len(text) <= limit else f"{text[:limit]}..."

File: frontend/src/lib/toolEvents.ts
import type { ChatMessage, ToolEventRecord } from "../types";

const FAILED_STATUSES = new Set(["failed", "error", "cancelled", "timeout"]);

export interface ToolSummary {
  label: string;
  tone: "normal" | "warning";
  details: ToolEventRecord[];
}

export function toolEventsFromMessage(message: ChatMessage): ToolEventRecord[] {
  return mergeToolEvents(
    message.toolEvents ?? [],
    readToolEvents(message.rawContent?.tool_events),
    readToolEvents(message.rawContent?.toolEvents),
    readToolEvents(message.rawContent?._toolEvents),
  );
}

export function mergeToolEvents(
  ...groups: ToolEventRecord[][]
): ToolEventRecord[] {
  const byKey = new Map<string, ToolEventRecord>();
  groups.flat().forEach((event, index) => {
    if (!event.toolName) return;
    const key =
      event.toolCallId || event.run_id || `${event.toolName}:${event.status || ""}:${index}`;
    byKey.set(key, { ...(byKey.get(key) ?? {}), ...event });
  });
  return Array.from(byKey.values());
}

export function summarizeToolEvents(
  events: ToolEventRecord[],
): ToolSummary | undefined {
  const completed = events.filter(
    (event) => String(event.status || "").toLowerCase() !== "running",
  );
  if (!completed.length) return undefined;
  const finalFailed = finalFailedToolEvents(completed);
  if (finalFailed.length) {
    return {
      label: `工具失败：${formatToolNames(finalFailed, 3).text}`,
      tone: "warning",
      details: completed,
    };
  }
  return {
    label: `调用：${formatToolNames(completed, 3).text}`,
    tone: "normal",
    details: completed,
  };
}

export function isFailedToolEvent(event: ToolEventRecord) {
  const status = String(event.status || "").toLowerCase();
  if (event.toolName === "terminal.wait_for" && status === "timeout") {
    return false;
  }
  if (FAILED_STATUSES.has(status)) return true;
  if (status === "succeeded" || status === "completed") return false;
  const exitCode = Number(event.exit_code);
  if (Number.isFinite(exitCode) && exitCode !== 0) return true;
  return Boolean(event.error);
}

export function toolEventDetailLines(event: ToolEventRecord): string[] {
  return [
    `tool: ${event.toolName}`,
    event.provider ? `provider: ${event.provider}` : "",
    event.run_id ? `run_id: ${event.run_id}` : "",
    `status: ${event.status || "unknown"}`,
    event.changed_files_count !== undefined
      ? `changed_files: ${event.changed_files_count}`
      : "",
    event.exit_code !== undefined ? `exit_code: ${event.exit_code}` : "",
    event.duration_ms !== undefined ? `duration_ms: ${event.duration_ms}` : "",
    event.stdout ? `stdout: ${shortText(event.stdout)}` : "",
    event.stderr ? `stderr: ${shortText(event.stderr)}` : "",
    event.error ? `error: ${shortText(event.error)}` : "",
    event.summary ? `summary: ${shortText(event.summary)}` : "",
  ].filter(Boolean);
}

function readToolEvents(value: unknown): ToolEventRecord[] {
  if (!Array.isArray(value)) return [];
  return value.map(normalizeToolEvent).filter((event) => Boolean(event.toolName));
}

function normalizeToolEvent(value: unknown): ToolEventRecord {
  if (!value || typeof value !== "object") return { toolName: "" };
  const record = value as Record<string, unknown>;
  return {
    toolName: stringValue(record.toolName ?? record.tool_name),
    toolCallId: stringValue(record.toolCallId ?? record.tool_call_id),
    run_id: stringValue(record.run_id ?? record.runId),
    provider: stringValue(record.provider),
    changed_files_count: primitiveValue(
      record.changed_files_count ?? record.changedFilesCount,
    ),
    status: stringValue(record.status),
    exit_code: primitiveValue(record.exit_code ?? record.exitCode),
    duration_ms: primitiveValue(record.duration_ms ?? record.durationMs),
    stdout: stringValue(record.stdout ?? record.stdout_tail ?? record.stdoutTail),
    stderr: stringValue(record.stderr ?? record.stderr_tail ?? record.stderrTail),
    summary: stringValue(record.summary ?? record.session_status ?? record.sessionStatus),
    error: stringValue(record.error ?? record.error_message),
    session_id: stringValue(record.session_id ?? record.sessionId),
    session_status: stringValue(record.session_status ?? record.sessionStatus),
    command: stringValue(record.command),
    cwd: stringValue(record.cwd),
  };
}

function finalFailedToolEvents(events: ToolEventRecord[]) {
  const latestByTool = new Map<string, ToolEventRecord>();
  events.forEach((event) => latestByTool.set(event.toolName, event));
  return Array.from(latestByTool.values()).filter(isFailedToolEvent);
}

function formatToolNames(events: ToolEventRecord[], maxTools: number) {
  const counts = new Map<string, number>();
  events.forEach((event) => {
    counts.set(event.toolName, (counts.get(event.toolName) ?? 0) + 1);
  });
  const names = Array.from(counts.entries());
  const visible = names.slice(0, maxTools).map(([name, count]) =>
    count > 1 ? `${name} ×${count}` : name,
  );
  const hiddenCount = names
    .slice(maxTools)
    .reduce((sum, [, count]) => sum + count, 0);
  return {
    text: hiddenCount
      ? `${visible.join(" · ")} · 等 ${hiddenCount} 项`
      : visible.join(" · "),
  };
}

function primitiveValue(value: unknown) {
  return typeof value === "string" || typeof value === "number"
    ? value
    : undefined;
}

function stringValue(value: unknown) {
  return typeof value === "string" ? value : "";
}

function shortText(value: string, limit = 160) {
  const text = value.replace(/\s+/g, " ").trim();
  return text.length > limit ? `${text.slice(0, limit)}...` : text;
}
